### ISSUE TITLE
Redundant $ lacks context

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/hlint#1024]
Given:

```
main = print (123 :: $(conT $ mkName $ "Int"))
```

I get the error "Redundant $" but the entire fragment I'm given is `$`. That's pretty hard to figure out what is going on. I think we probably need to revert #950 - likely having an "I'm an IDE flag" to say that the span is super important but the context less-so.

**Comment by ndmitchell** [comment]
CC @zliu41

**Comment by zliu41** [comment]
No objection to reverting - the IDE aspect of this particular hint is not very important (to me anyway) since I never keep redundant "$"s, so if I see it I just fix it.

On the other hand, both the LHS and RHS of the "$" may be quite large and may have additional "$"s, so with the old span, one may sometimes have to squint at the "Found" and "Perhaps" to find the "$". The new span at least says exactly where the redundant "$" is.

**Comment by ndmitchell** [comment]
I'm thinking that leaving the position as just the position of `$`, but making the from be the whole LHS/RHS. I agree that for big LHS/RHS it's ugly, but that means IDE users (who have context) get the right answer, whereas console users know which $ is being complained about. That seems like the best of both worlds?

**Comment by ndmitchell** [comment]
Done. I now get:

```
Sample.hs:1:39: Suggestion: Redundant $
Found:
  mkName $ "Int"
Perhaps:
  mkName "Int"
```

So context for the command line, but a position that only refers to a single character for the IDE.

**Comment by zliu41** [comment]
> I'm thinking that leaving the position as just the position of $, but making the from be the whole LHS/RHS.

Great idea, I think we can do the same for redundant brackets.

**Comment by ndmitchell** [comment]
>  I think we can do the same for redundant brackets.

Awesome! Yes, this looks like it could well become a standard trick.

**Comment by ndmitchell** [comment]
Just going to hack up the redundant brackets now.

**Comment by ndmitchell** [comment]
Brackets trimmed down in 90dde61ab2039597709e20f444a369d3a1f791b5 - was pretty easy now we have a general technique.
